cmd/guacone/cmd/eol.go

```diff
@@ -0,0 +1,281 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"os/signal"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/Khan/genqlient/graphql"
+	"github.com/guacsec/guac/pkg/assembler/clients/generated"
+	"github.com/guacsec/guac/pkg/certifier/certify"
+	"github.com/guacsec/guac/pkg/certifier/components/root_package"
+	"github.com/guacsec/guac/pkg/certifier/eol"
+	"github.com/guacsec/guac/pkg/cli"
+	csub_client "github.com/guacsec/guac/pkg/collectsub/client"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/ingestor"
+	"github.com/guacsec/guac/pkg/logging"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+const (
+	eolQuerySize = 1000
+)
+
+type eolOptions struct {
+	graphqlEndpoint   string
+	headerFile        string
+	poll              bool
+	csubClientOptions csub_client.CsubClientOptions
+	interval          time.Duration
+	addedLatency      *time.Duration
+	batchSize         int
+	lastScan          *int
+}
+
+var eolCmd = &cobra.Command{
+	Use:   "eol [flags]",
+	Short: "runs the End of Life (EOL) certifier",
+	Run: func(cmd *cobra.Command, args []string) {
+		opts, err := validateEOLFlags(
+			viper.GetString("gql-addr"),
+			viper.GetString("header-file"),
+			viper.GetString("interval"),
+			viper.GetString("csub-addr"),
+			viper.GetBool("poll"),
+			viper.GetBool("csub-tls"),
+			viper.GetBool("csub-tls-skip-verify"),
+			viper.GetString("certifier-latency"),
+			viper.GetInt("certifier-batch-size"),
+			viper.GetInt("last-scan"),
+		)
+		if err != nil {
+			fmt.Printf("unable to validate flags: %v\n", err)
+			_ = cmd.Help()
+			os.Exit(1)
+		}
+
+		ctx := logging.WithLogger(context.Background())
+		logger := logging.FromContext(ctx)
+		transport := cli.HTTPHeaderTransport(ctx, opts.headerFile, http.DefaultTransport)
+
+		if err := certify.RegisterCertifier(eol.NewEOLCertifier, eol.EOLCollector); err != nil {
+			logger.Fatalf("unable to register certifier: %v", err)
+		}
+
+		// initialize collectsub client
+		csubClient, err := csub_client.NewClient(opts.csubClientOptions)
+		if err != nil {
+			logger.Infof("collectsub client initialization failed, this ingestion will not pull in any additional data through the collectsub service: %v", err)
+			csubClient = nil
+		} else {
+			defer csubClient.Close()
+		}
+
+		httpClient := http.Client{Transport: transport}
+		gqlclient := graphql.NewClient(opts.graphqlEndpoint, &httpClient)
+		packageQuery := root_package.NewPackageQuery(gqlclient, generated.QueryTypeEol, opts.batchSize, eolQuerySize, opts.addedLatency, opts.lastScan)
```

Comment:
The query will need to be updated to include the new query type: 

For the ENT Backend (main optimized backend) this will be here: https://github.com/guacsec/guac/blob/main/pkg/assembler/backends/ent/backend/search.go#L127

And for Keyvalue (this backend is just used for testing and not very optimized) here: https://github.com/guacsec/guac/blob/main/pkg/assembler/backends/keyvalue/search.go#L313

Reply:
Let me know if you have questions around the implementation, happy to help!

Reply:
That is good to know, I thought I might be missing something here, thanks Parth! I might reach out later on slack.